src/lib/reasoners/satml.ml

```diff
@@ -1318,15 +1388,34 @@ module Make (Th : Theory.S) : SAT_ML with type th = Th.t = struct
       | [] -> assert false
       | [(fuip : Atom.atom)] ->
         fuip.var.vpremise <- history;
+        (* Note: there could potentially be semantic values that are no longer
+           valid in the union-find at level 0 contained in [fuip] if [fuip] is
+           a semantic literal, which could cause crashes -- although it is not
+           clear to me that it can really happen since if we learn it at level
+           0 it must be explainable by facts at level 0. *)
         enqueue env fuip 0 None
       | fuip :: _ ->
         let name = Atom.fresh_lname () in
         let lclause =
           Atom.make_clause name learnt vraie_form true history
         in
-        Vec.push env.learnts lclause;
-        attach_clause env lclause;
-        clause_bump_activity env lclause;
+        (* Do not learn clauses involving semantic literals (but still record
+           them as a reason for the propagated literal), otherwise we could
+           learn stuff about these semantic literals in a scope where they are
+           no longer valid.
```

Comment:
I'm not sure to understand this comment. Do you avoid semantic literals here for sake of soundness or efficiency? From my understanding, if we learn something in the CDCL framework, this fact have to be entailed by the problem. It may be irrelevant after backtracking but still correct. So I guess you meant we should avoid learning semantic facts because the case-split mechanism could produce a lot of them and we don't want to increase too much the size of the clause database?

Reply:
Clauses are learned at level 0 but semantic literals can includes terms created by the Shostak module that depend on assertions at greater levels, and hence do not exist (in the union-find etc.) at level 0 (what I mean by "no longer valid" here). I have updated and clarified the comment.